proposals/0215-additive-arithmetic-protocol.md

```diff
@@ -0,0 +1,209 @@
+
+Discussion thread: [Should Numeric not refine ExpressibleByIntegerLiteral](https://forums.swift.org/t/should-numeric-not-refine-expressiblebyintegerliteral/15106)
+
+## Motivation
+
+The  `Numeric`  protocol today refines  `ExpressibleByIntegerLiteral`  and defines all arithmetic operators. The design makes it easy for scalar types to adopt arithmetic operators, but makes it hard for vector types to adopt arithmetic operators by conforming to this protocol.
+
+What's wrong with `Numeric`? Assuming that we need to conform to `Numeric` to get basic arithmetic operators and generic algorithms, we have three problems.
+
+### 1. Vectors conforming to `Numeric` would be mathematically incorrect.
+
+`Numeric` roughly corresponds to a [ring](https://en.wikipedia.org/wiki/Ring_(mathematics)). Vector spaces are not rings. Multiplication is not defined between vectors. Requirements `*` and `*=` below would make vector types inconsistent with the mathematical definition.
+
+```
+static func * (lhs: Self, rhs: Self) -> Self
+static func *= (lhs: inout Self, rhs: Self)
+```
+
+### 2. Literal conversion is undefined for dynamically shaped vectors.
+
+Vectors can be dynamically shaped, in which case the the shape needs to be provided when we initialize a vector from a scalar. Dynamically shaped vector types often have an initializer `init(repeating:shape:)`.
+
+Conforming to `Numeric` requires a conformance to `ExpressibleByIntegerLiteral`, which requires `init(integerLiteral:)`. However, the conversion from a scalar to a dynamically shaped vector is not defined when there is no given shape.
+
+```swift
+struct Vector<Scalar: Numeric>: Numeric {
+  // Okay!
+  init(repeating: Scalar, shape: [Int]) { ... }
+
+  // What's the shape?
+  init(integerLiteral: Int)
+}
+```
+
+### 3. Common operator overloading causes type checking ambiguity.
+
+Vector types mathematically represent [vector spaces](https://en.wikipedia.org/wiki/Vector_space). Vectors by definition do not have multiplication between each other, but they come with scalar multiplication.
+
+```swift
+static func * (lhs: Vector, rhs: Scalar) -> Vector { ... }
+```
+
+By established convention in numerical computing communities such as machine learning, many libraries define a multiplication operator `*` between vectors as element-wise multiplication. Given that scalar multiplication has to exist by definition, element-wise multiplication and scalar multiplication would overload the `*` operator.
+
+```swift
+static func * (lhs: Vector, rhs: Vector) -> Vector { ... }
+static func * (lhs: Vector, rhs: Scalar) -> Vector { ... }
+```
+
+This compiles, but does not work in practice. The following trivial use case would fail to compile, because literal  `1`  can be implicitly converted to both a  `Scalar`  and a  `Vector` , and  `*`  is overloaded for both  `Vector`  and  `Scalar` .
+
+```swift
+let x = Vector<Int>(...)
+x * 1 // Ambiguous! Can be both `x * Vector(integerLiteral: 1)` and `x * (1 as Int)`.
```

Comment:
I meant to note that this isn't ambiguous, since the type checker will always first try to use an literal's default type (`Int` for integer literals). If there's a use that supports that, the default wins. When there's nothing available using the default, then there's an ambiguity, so making this example use a `Vector<Double>` would be a better illustration. There's no precedence in that case between `Vector<Double>` and `Vector<Double>.Scalar` (i.e. `Double`), so there's an ambiguity.